linked_list/linked_list.py

```diff
@@ -1,82 +1,174 @@
 
-# Defines a node in the singly linked list
+# Defines a node in the doubly linked list
 class Node:
 
-    def __init__(self, value, next_node = None):
+    def __init__(self, value, next_node = None, previous_node = None):
         self.value = value
         self.next = next_node
+        self.prev = previous_node
 
-# Defines the singly linked list
+# Defines the doubly linked list
 class LinkedList:
     def __init__(self):
-      self.head = None # keep the head private. Not accessible outside this class
+        self.__head = None # keep the head private. Not accessible outside this class
+        self.__tail = None
 
     # returns the value in the first node
     # returns None if the list is empty
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(1)
+    # Space Complexity: O(1)  
     def get_first(self):
-        pass
+        if self.__head == None:
+            return
+
+        return self.__head.value
 
 
     # method to add a new node with the specific data value in the linked list
     # insert the new node at the beginning of the linked list
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(1)
+    # Space Complexity: O(1)
     def add_first(self, value):
-        pass
+        node = Node(value)
+
+        if self.__head == None: 
+            self.__head = self.__tail = Node(value)
+        else: 
+            temp = self.__head
+            self.__head.prev = node
+            self.__head = node
+            self.__head.next = temp
 
+    
     # method to find if the linked list contains a node with specified value
     # returns true if found, false otherwise
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity: O(1)
     def search(self, value):
-        pass
+        if self.__head == None:
+            return False
+
+        current = self.__head
+
+        if current.value == value: 
+            return True
+
+        while current.next: 
+            current = current.next
+            if current.value == value:
+                return True
+        
+        return False
 
     # method that returns the length of the singly linked list
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity: O(1)
     def length(self):

```

Comment:
👍   You could also do the `length` method by maintaining a number field called length and add one to it when you add a new node to the list and subtract one when you remove a node from the list.